docs/testing.md

```diff
@@ -0,0 +1,138 @@
+# Testing BMI
+
+Every Developer will need to test his/her feature or make sure that he/she
+did not break anything. So the test cases are situated in
+
+```
+/tests/
+```
+
+## Directory Structure
+
+This folder follows the directory structure of the project with additional
+folders like for example operation.py's test cases are situated in
+```
+/tests/einstein/operations/unit/...
+/tests/einstein/operations/integration/...
+```
+
+Every file must have a folder for itself which has additional folders called
+'unit' and 'integration'
+
+All the unit tests should be in unit folder and all integration tests should
+be in integration folder for that file
+
+All of these folders are python packages so they should have a \_\_init\_\_.py
+
+## Writing Tests
+
+Create a file in the correct location of the directory structure and make sure
+the name conforms to the regex 'test_\*.py'
+
+We are using python's unittest as our testing framework for now.
+
+So the file can have any number of test cases as long as it makes sense they
+belong together.
+
+Each test case is represented by a class and should extend unittest.TestCase
+
+This is the typical structure of a test case
+
+```
+from ims.common.log import trace
+
+class TestSample(unittest.TestCase):
+
+  @trace
+  def setUp(self):
+       pass
+
+  def runTest(self):
+       pass
+
+  def tearDown(self):
+       pass
+ ```
+ 
+* Each class's name should start with Test
+* Each setUp method should use the trace decorator
+* setUp method is supposed to do stuff which is required before the actual test
+is run like adding db entries, etc.
+* runTest should do the actual test and assertions to make sure everything went
+ well.
+* tearDown is supposed to do cleanup most commonly undo what was done in setUp
+like delete db entries, etc.
+ 
+Each test case should be independent of running order, the setUp and tearDown
+methods should handle this.
+ 
+It is fine if a failed test case causes subsequent tests to fail.
+ 
+Each test can use the test variables in the config file
+ 
+```
+import ims.common.config as config
+ 
+cfg = config.get()
+ 
+IMAGE_NAME = cfg.tests.image_name
+```
+The above example is the way to access these variables
+ 
+Any key value pair declared in the [tests] section of the config will be
+available in cfg.tests variable.
+ 
+The current values that are available in our CI are
+
+* picasso_url
+* correct_hil_username
+* correct_hil_password
+* incorrect_hil_password
+* node_name
+* nic
+* project
+* network
+* exist_img_name
+* new_snap_name
+* not_exist_snap_name
+* not_exist_img_name
+
+Let us know if your test cases need additional variables
+
+## Running Tests
+
+We made a testing framework that collects and runs tests. So follow these steps
+
+* Write a proper bmiconfig.cfg
+* Set the variable BMI_CONFIG
+```
+export BMI_CONFIG=<path>
+```
+* Some tests may require Picasso and Einstein to be running so do that if
+required before proceeding.
```

Comment:
If a test requires Picasso and Einstein to be running, then the test should start those under `SetUp`. That way we can run these tests in Travis. What do you say?